### PR TITLE
Add new Origin Trials API

### PR DESCRIPTION
**Pull request by DanielRyanSmith**
This PR adds the new Origin Trials API that will be used to interact with the OT API on the server side. A new endpoint is also created to query the OT API for origin trial data.

### CODE REVIEW

File: api/origin_trials_api.py
Comment (jcscottiii) [inline api/origin_trials_api.py:26]: If we are planning on adding more requests besides the GET, we could move all the OT client requests logic to a new file in `frameworks`. Then we can cleanly separate OT client code from the server handler code.

File: api/origin_trials_api.py
Comment (jcscottiii) [inline api/origin_trials_api.py:2]:
```suggestion
# Copyright 2023 Google Inc.
```

File: api/origin_trials_api.py
Comment (jcscottiii) [inline api/origin_trials_api.py:31]: If the key exists, we should wrap the request and the parsing in a try-except. In the event something goes wrong, log the exception internally and send back a generic error for the frontend to display.

Comment (DanielRyanSmith) [reply api/origin_trials_api.py]: Yes, agreed - I've added a try/except block with the approach you mentioned 👍

File: api/origin_trials_api.py
Comment (DanielRyanSmith) [reply api/origin_trials_api.py]: I want to make sure I'm understanding correctly here - We want to separate code for handling client-side requests vs. server side requests like automated trial creation, is that right? If so, would it be better to keep the server handler code in `framework` instead, and keep the client API request handlers in this file? I'm not certain of the best organization here

File: api/origin_trials_api.py
Comment (jrobbins) [inline api/origin_trials_api.py:39]: Can this be `except HTTPException` or something else a little more specific?  The problem with a catch-all `except:` is that it can mask programming errors.
https://google.github.io/styleguide/pyguide.html#244-decision

File: api/origin_trials_api.py
Comment (jrobbins) [inline api/origin_trials_api.py:38]: Unit tests would help document the expected type of `trials`.

**Review by jrobbins — CHANGES_REQUESTED** [review]

File: api/origin_trials_api.py
Comment (jcscottiii) [inline api/origin_trials_api.py:38]: You should print a log with an exception level as well. (So that we get an email alert of some sorts)

File: api/origin_trials_api.py
Comment (jcscottiii) [reply api/origin_trials_api.py]: api/origin_trials_api.py should handle the incoming request to the ChromeStatus API and returning an appropriate response.

framework/origin_trials_client.py should handle sending any requests to the OT API.

The workflow would be like this
1. incoming request to the Chromestatus API which is handled by `api/origin_trials_api.py`
2. `api/origin_trials_api.py` parses the request and validates any passed in data (Not for this current GET method but you can imagine in the POST method)
3. `api/origin_trials_api.py` calls `framework/origin_trials_client.py` to do some operation
4. `framework/origin_trials_client.py` calls framework/secrets to get the OT secret and then makes the appropriate call the OT API. (In the future, `framework/origin_trials_client.py`, could raise certain custom exceptions which could be caught by `api/origin_trials_api.py` to let it know which http error to return)
5. `api/origin_trials_api.py` gets the returned value and sends the response back

File: api/origin_trials_api.py
Comment (jcscottiii) [inline api/origin_trials_api.py:40]: I know it's low likelihood since we work on both OT API and ChromeStatus API now, but in the event the OT API ever changes and `trials` is missing, this line will throw an uncaught exception.

Technically, we should check if there's a trials key. If not, throw a 500 (since the server did not receive expected data to send back to the end user).

Comment (jcscottiii) [reply api/origin_trials_api.py:40]: Also, what is the exact data that we need to return back to the frontend? We should be explicit about the shape of the data we want to send back. That will prevent accidentally oversharing in case a field is added on the OT API side. Also, it will help out on the frontend to have the expected data response schema somewhere in this code.

**Review by jcscottiii — CHANGES_REQUESTED** [review]

File: api/origin_trials_api.py
Comment (DanielRyanSmith) [reply api/origin_trials_api.py:40]: > Technically, we should check if there's a trials key. If not, throw a 500 (since the server did not receive expected data to send back to the end user).

Gotcha - I've added a check for this.

> Also, it will help out on the frontend to have the expected data response schema somewhere in this code.

Is this where OpenAPI comes in? I know we spoke about it, but I haven't familiarized myself with it yet. 😅 If this is the appropriate time, I'm willing to do so

File: api/origin_trials_api.py
Comment (DanielRyanSmith) [reply api/origin_trials_api.py]: I see, that makes sense - I'll try and implement this structure

File: api/origin_trials_api.py
Comment (DanielRyanSmith) [reply api/origin_trials_api.py]: I went to go do this, but it seems logging is already implemented in the `self.abort` call

https://github.com/GoogleChrome/chromium-dashboard/blob/main/framework/basehandlers.py#L67-L71

File: api/origin_trials_api.py
Comment (jcscottiii) [reply api/origin_trials_api.py]: A dataclass will be sufficient. I think the plan is to use OpenAPI for the v1 of the API.

File: api/origin_trials_api.py
Comment (jcscottiii) [reply api/origin_trials_api.py]: Oh ok! Thanks for pointing that out.

File: internals/data_types.py
Comment (jcscottiii) [inline internals/data_types.py:276]: Will the frontend need to know about the `originTrialFeatureName` field?

File: framework/origin_trials_client.py
Comment (jcscottiii) [inline framework/origin_trials_client.py:31]: The exceptions should be documented in this method docstring  https://google.github.io/styleguide/pyguide.html#244-decision

The two exceptions will help origin_trials_api.py know which message to put in the abort.

**Review by jcscottiii — CHANGES_REQUESTED** [review]
The frameworks/origin_trials_client.py should not have knowledge of the type of http status codes or message to return to the frontend. Instead of returning a tuple of value and error, you need to raise an error.

https://google.github.io/styleguide/pyguide.html#244-decision

Then in api/origin_trials_api.py, you can do things like:

```py
try:
  trials_list = origin_trials_client.get_trials_list()
except ExceptionType1:
  self.abort(500, 'Error obtaining origin trial data from API')
except ExceptionType2:
  self.abort(500, 'Malformed response from origin trials API')
```

Regarding creating your own Exception types vs re-raising existing ones: I'm okay re-raising the existing ones for now. We can always refactor that later on.

File: internals/data_types.py
Comment (DanielRyanSmith) [reply internals/data_types.py:273]: Probably not needed, but I'm not certain - I've just removed it for now

File: internals/data_types.py
Comment (DanielRyanSmith) [reply internals/data_types.py:279]: Until trial creation is done on ChromeStatus, this is one of the only ways to identify which trials belong to which ChromeStatus features, so we'll use this on the client side, at least for a little while

File: framework/origin_trials_client.py
Comment (DanielRyanSmith) [reply framework/origin_trials_client.py:42]: I want to make sure I'm doing this right - should I not be passing the error info in the return value and instead be returning an actual error? Or am I raising the error in the `get_trials_list()` function and then catching it in the `origin_trials_api.do_get()` to raise it again there?

Comment (jcscottiii) [reply framework/origin_trials_client.py:42]: You'll do this part `Or am I raising the error in the get_trials_list() function and then catching it in the origin_trials_api.do_get()`. 

Then in do_get(), it will catch that specific exception and call the right abort:

```py
try:
  trials_list = origin_trials_client.get_trials_list()
except ExceptionType1: # the http one
  self.abort(500, 'Error obtaining origin trial data from API')
except ExceptionType2: # The valueerror one
  self.abort(500, 'Malformed response from origin trials API')
```

File: internals/data_types.py
Comment (jcscottiii) [reply internals/data_types.py:279]: Makes sense! Thanks!

File: internals/data_types.py
Comment (jcscottiii) [reply internals/data_types.py:276]: Wouldn't we want to use the `id` field for that purpose instead?

Comment (DanielRyanSmith) [reply internals/data_types.py:276]: Yes, but we cannot know the OT ID since we do not have access to that until trial creation is handled on Chromestatus, or we ask feature owners to supply the OT ID in Chromestatus. We ask feature owners to supply this feature name field, and with this + the chromestatus URL, we can begin to associate origin trials to their respective feature stage in Chromestatus.

This is not an implementation we want to use permanently, but it would make it possible with what we currently have to link OTs with their feature, and then when trial creation is handled in Chromestatus, we can save the trial ID directly to make this association.

I am willing to talk more about this if you would like to talk about the efficacy of this approach 🙂 I won't deny that it's half-baked 😅

Comment (jcscottiii) [reply internals/data_types.py:276]: No worries! Thanks for the explanation. It helped. I get what you're going for now. Let's go with what you have and we can revisit in the future if needed.

File: framework/origin_trials_client.py
Comment (jcscottiii) [inline framework/origin_trials_client.py:31]:
```suggestion
  """Get a list of all origin trials.
  
  Raises:
    requests.exceptions.RequestException: If the request fails to connect or the HTTP
      status code is not successful.
    KeyError: If the response from the OT API is not in the expected format.
  """
```

File: framework/origin_trials_client.py
Comment (jcscottiii) [inline framework/origin_trials_client.py:37]: You should still have the try-except for the request.

File: framework/origin_trials_client.py
Comment (jcscottiii) [inline framework/origin_trials_client.py:40]:
```suggestion
  try:
    response = requests.get(
        f'{settings.OT_API_URL}/v1/trials',
        params={'prettyPrint': 'false', 'key': key})
    response.raise_for_status()
  except requests.exceptions.RequestException as e:
    logging.exception("Failed to get response from OT")
    raise e
```

This will give us a friendly log statement before doing a stack trace print to the logs (which should send an email to us)

You will need to `import logging`

File: framework/origin_trials_client.py
Comment (jcscottiii) [inline framework/origin_trials_client.py:27]:
```suggestion
```

File: framework/origin_trials_client.py
Comment (jcscottiii) [inline framework/origin_trials_client.py:30]:
```suggestion
def get_trials_list() -> list[dict[str, Any]]:
```

File: framework/origin_trials_client.py
Comment (jcscottiii) [inline framework/origin_trials_client.py:35]:
```suggestion
    return []
```

File: framework/origin_trials_client.py
Comment (jcscottiii) [inline framework/origin_trials_client.py:47]:
```suggestion
  return trials_list
```

File: framework/origin_trials_client.py
Comment (jcscottiii) [inline framework/origin_trials_client.py:46]:
```suggestion
  try:
    trials_list = [asdict(OriginTrialInfo(api_trial))
                   for api_trial in response_json['trials']
                   if api_trial['isPublic']]
  except KeyValue as e:
    logging.exception("Received badly formatted data from OT API")
    raise e
```

File: api/origin_trials_api.py
Comment (jcscottiii) [inline api/origin_trials_api.py:29]:
```suggestion
    
    Raises:
      RequestException: If an error occurred interacting with the OT API.
      KeyError: If the response from the OT API is not in the expected format.
```

This doc string should be for the get_trials_list method as the exceptions originate there. Will make a suggestion there. Remove the raises section here.

**Review by jcscottiii — CHANGES_REQUESTED** [review]

File: framework/origin_trials_client.py
Comment (DanielRyanSmith) [reply framework/origin_trials_client.py]: Ah, I wasn't sure if I was supposed to leave the try/except block in here along with the one in the `origin_trials_api` function

File: framework/origin_trials_client.py
Comment (DanielRyanSmith) [reply framework/origin_trials_client.py]: Sorry, I had thought I removed these, as well as the error type hinting, but it's not showing up here... Maybe it was with an abandoned change prior. Either way, I should have checked before sending for a second review!

File: framework/origin_trials_client.py
Comment (DanielRyanSmith) [reply framework/origin_trials_client.py]: Ah, I wasn't sure if I was supposed to leave the try/except block in here along with the one in the `origin_trials_api` function

File: framework/origin_trials_client.py
Comment (DanielRyanSmith) [reply framework/origin_trials_client.py]: Sorry, I remember removing the code for the error returns and the type hinting, but it doesn't show up here. It might have been part of an abandoned change. Either way, I should have caught it before sending it over for re-review!

**Review by jcscottiii — APPROVED** [review]

**Pull request merged**
